Add option to validate transaction signatures in the background

bboston7 opened the pull request:
This change adds an experimental config option to validate transaction signatures in the background when receiving transactions over the network.

This PR may need updating to integrate with #4692.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md) (testing in progress)

marta-lokhova commented on src/overlay/Peer.cpp:
```diff
@@ -49,6 +51,38 @@ static std::string const AUTH_ACTION_QUEUE = "AUTH";
+    auto const& hash = tx->getContentsHash();
+    auto const& signatures = txbridge::getSignatures(tx->getEnvelope());
+    auto const snapshot = app.getOverlayThreadSnapshot();
+    BucketSnapshotState bucketSnapshot(snapshot);
```
For consistency, please use a wrapper for read-only ledger snapshots `LedgerSnapshot`

bboston7 replied:
Done!

marta-lokhova commented on src/overlay/Peer.cpp:
```diff
@@ -49,6 +51,38 @@ static std::string const AUTH_ACTION_QUEUE = "AUTH";
+    // Check signature, which will add the result to the signature cache.
+    tx->checkSignature(
+        signatureChecker, sourceAccount,
+        sourceAccount.current().data.account().thresholds[THRESHOLD_HIGH]);
```
Is this correct? In TransactionFrame::commonValid, we pass `sourceAccount.current().data.account().thresholds[THRESHOLD_LOW]`

bboston7 replied:
I don't think LOW is right either. I originally had this set to LOW based on that exact call in `TransactionFrame::commonValid`, but I was still seeing some cache misses. Then I discovered that [`OperationFrame` calls into `TransactionFrame::checkSignature`](https://github.com/stellar/stellar-core/blob/3b77d871f2f3a549a9e0daffd2ceb993477245ef/src/transactions/OperationFrame.cpp#L186) with a default threshold of MEDIUM. It looks like Payment operations don't change this threshold, for example.

So doesn't it ultimately depend on the operations in the transaction? I set this at HIGH figuring it would cover all of the bases, and there's plenty of room on the background thread anyway. I'm open to lowering to MEDIUM to cover the common case, but I think LOW might miss a bit (unless I'm misunderstanding how this works).

I haven't looked at all of the `TransactionFrameBase` implementations, so I'm not sure if it's feasible, but perhaps we could add a virtual function to `TransactionFrameBase` to get any inner `OperationFrame`s, kind of like we have in `TransactionFrame`. Then we would be able to get the max threshold in any `OperationFrame` and pass that off to the signature checker.

What do you think?

marta-lokhova replied:
> I originally had this set to LOW based on that exact call in TransactionFrame::commonValid, but I was still seeing some cache misses

Interesting, that seems surprising. Default accounts have all thresholds set to 0, and master key weight set to 1. So for the setup we were testing it shouldn't matter what threshold value we pass, as we should always verify the signature of the source account and exit the loop. 

The change to `THRESHOLD_HIGH` should only matter for multi-sig situations with different threshold and signer weights. Initially I thought we'll just cover the most common case with a single signer, but if it's safe to check all signers in the background, I agree we could do that. We're not using the result of checkSignature, but simply populating the cache. @sisuresh what do you think?

sisuresh replied:
The only side affect of this call is populating the signature cache right? I don't see any safety issues with increasing the threshold so additional signatures are checked. 

Keep in mind that we don't enforce an order between the thresholds, so someone could set HIGH less than MEDIUM on their account. You could in theory still see some cache misses if an account had a weird setup like this. If you wanted to go over every signature that will be verified later, you could just pass in UINT8_MAX for the threshold, but I'd expect this to be an edge case that doesn't need to be optimized.

bboston7 replied:
> Interesting, that seems surprising. Default accounts have all thresholds set to 0, and master key weight set to 1. So for the setup we were testing it shouldn't matter what threshold value we pass, as we should always verify the signature of the source account and exit the loop.

To be clear, I saw this on pubnet, not necessarily our test setup. LOW might be sufficient for our max TPS tests, but for "real world" use cases I saw an impact on number of main thread cache misses when running a watcher on pubnet.

> The only side affect of this call is populating the signature cache right?
Correct

> Keep in mind that we don't enforce an order between the thresholds, so someone could set HIGH less than MEDIUM on their account. You could in theory still see some cache misses if an account had a weird setup like this. If you wanted to go over every signature that will be verified later, you could just pass in UINT8_MAX for the threshold, but I'd expect this to be an edge case that doesn't need to be optimized.

That is interesting.

marta-lokhova replied:
Would be great to add a comment explaining protocol safety assumptions with these caching changes (why it's ok to populate this cache before validation), and also explain why passing a high threshold is safe here.

bboston7 replied:
I added a comment here

SirTyson commented on src/main/AppConnector.h:
```diff
@@ -62,5 +62,9 @@ class AppConnector
+
+    // Get a snapshot of ledger state for use by the overlay thread only. Must
+    // only be called from the overlay thread.
+    SearchableSnapshotConstPtr getOverlayThreadSnapshot();
```
I'm not sure if we want to maintain a persistent, self updating BucketListSnapshot for the overlay thread, as self-updating snapshots have bitten us before. We should either just create a new snapshot every time, or if we do want to persist a snapshot for performance reasons, maybe we should manually make calls to update as necessary.

bboston7 replied:
That's fair. I assume that would mean calling `copySearchableLiveBucketListSnapshot` on every transaction though. Is that performant enough to call 1000+ times per second? This is an honest question; I can't tell whether `copySearchableLiveBucketrListSnapshot` makes a copy of the snapshot itself, or if it's making a copy of a pointer to a pre-existing immutable snapshot (fast), or something else.

SirTyson replied:
It's fairly cheap, it's just a pointer copy + opening up some new file stream. 1000+ times per seconds probably means the file streams will be expensive though. In that case I'd keep the single snapshot, but remove the auto-updating logic from the getter and just to it explicitly after getting the snapshot. That way if we have some logic in a future flow that requires a consistent snapshot between several loads, we won't run into this potential footgun.

bboston7 replied:
Got it. I kept the snapshot, but removed the auto-updating. Now I just update it before using it.

marta-lokhova commented on src/overlay/Peer.cpp:
```diff
@@ -967,20 +1002,64 @@ Peer::recvAuthenticatedMessage(AuthenticatedMessage&& msg)
+        // Verify signature in the background
+        mAppConnector.postOnOverlayThread(
+            [self = shared_from_this(), msgTracker = std::move(msgTracker),
+             queueName = std::move(queueName), type]() mutable {
```
Is this correct? queueName is determined based on the state of the peer at the time of message processing: since here we bounce to a different thread, the state may change, causing us to post the message to a wrong queue. Ideally, to avoid footguns, we should keep this block synchronous. I recall you experimented with unlocking + verifying signature + locking, but I don't recall why we moved away from that approach. Any thoughts? 

Another thing: since we still process transactions individually, we can just do signature verification synchronously while holding the lock. Given that we don't do batches in prod, that shouldn't cause too much extra blocking. We can tune the locking strategy in the future when we enable batching by default.

marta-lokhova replied:
Thinking about this more: I'd prefer to go with the simplest option to just do verification synchronously with the lock, and tune locking later if we find it to be a problem, or when we decide to enable batching by default. Open to discussion on this though.

marta-lokhova replied:
Also, @SirTyson recently removed a problematic assert that was holding the lock on main, I wonder if we observe the same locking issue now that this assert is removed?

bboston7 replied:
>  I recall you experimented with unlocking + verifying signature + locking, but I don't recall why we moved away from that approach.

I couldn't find a nice way to do this that I was 100% confident was correct. [TCPPeer acquires the lock a few functions up the stack](https://github.com/stellar/stellar-core/blob/63fe672849c517c8c9cce72a6e65090371138eb4/src/overlay/TCPPeer.cpp#L494), so it's hard to drop the lock just for signature verification without posting back to the overlay thread.

> Also, @SirTyson recently removed a problematic assert that was holding the lock on main, I wonder if we observe the same locking issue now that this assert is removed?

That should help reduce main thread contention.

> I'd prefer to go with the simplest option to just do verification synchronously with the lock, and tune locking later if we find it to be a problem, or when we decide to enable batching by default. Open to discussion on this though.

I agree it doesn't make much of a difference without batching. For non-batched transactions, the perf tests came back similarly between the synchronous approach and the async approach. I'm happy to use the simpler, synchronous approach here until batching is the default behavior (if the issue persists in whatever implementation we finalize).

bboston7 replied:
I switched this to a synchronous approach

bboston7 commented:
In addition to addressing the above comments, I also rebased on the release branch and added support for batched transactions. I moved some of the control flow away from `recvAuthenticatedMessage` and into the `CapacityTrackedMessage` constructor to better line up with the batching change.

SirTyson commented on src/overlay/Peer.cpp:
```diff
@@ -54,14 +54,17 @@ using namespace soci;
+// the process. This function does nothing if background signature verification
+// is disabled and/or the current thread is not the overlay thread.
 void
 checkTransactionSignature(AppConnector& app, TransactionFrameBaseConstPtr tx)
```
Given that this function can no-op, maybe a more descriptive name would be "populateSignatureCache". I also am not a huge fan of dictating logic based on the current thread (beyond asserts) and it seems like a footgun to rely on thread PIDs. Is it possible to keep this check as an assert and move the conditional logic to the caller?

bboston7 replied:
I renamed this to `populateSignatureCheck` and switched the check back to an assert. The check now occurs in the `CapacityTrackedMessage` constructor and relies on the `useBackgroundThread` function rather than checking thread ids.

SirTyson reviewed (approved):
Overall looks good to me, just a couple of comments on small things.

marta-lokhova reviewed (approved):
LGTM!

bboston7 commented on src/overlay/Peer.cpp:
```diff
@@ -49,6 +51,54 @@ static std::string const AUTH_ACTION_QUEUE = "AUTH";
+checkTransactionSignature(AppConnector& app, TransactionFrameBaseConstPtr tx)
+{
+    ZoneScoped;
+    if (!(app.getConfig().EXPERIMENTAL_BACKGROUND_TX_SIG_VERIFICATION &&
```
I wound up implementing [@SirTyson's suggestion](https://github.com/stellar/stellar-core/pull/4699#discussion_r2067398539), renaming this to `populateSignatureCache` in the process.

bboston7 commented:
I've addressed all the comments, rebased on the release branch, and squashed down to a single commit, so I think this is ready to merge!

SirTyson reviewed (approved):
LGTM!
